rapids_logger/logger.hpp.in

```diff
@@ -0,0 +1,427 @@
+  {
+    log(level_enum::debug, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the INFO level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void info(std::string const& format, Args&&... args)
+  {
+    log(level_enum::info, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the WARN level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void warn(std::string const& format, Args&&... args)
+  {
+    log(level_enum::warn, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the ERROR level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void error(std::string const& format, Args&&... args)
+  {
+    log(level_enum::error, format, std::forward<Args>(args)...);
+  }
+
+  /**
+   * @brief Log a message at the CRITICAL level.
+   *
+   * @param format The format string
+   * @param args The format arguments
+   */
+  template <typename... Args>
+  void critical(std::string const& format, Args&&... args)
+  {
+    log(level_enum::critical, format, std::forward<Args>(args)...);
+  }
+
+  // Everything below here is conditionally compiled based on whether logging is supported.
+#ifdef @_RAPIDS_LOGGER_MACRO_PREFIX@_SUPPORTS_LOGGING
+  /**
+   * @brief Construct a new logger object
+   *
+   * @param name The name of the logger
+   * @param filename The name of the log file
+   */
+  logger(std::string name, std::string filename);
+
+  /**
+   * @brief Construct a new logger object
+   *
+   * @param name The name of the logger
+   * @param stream The stream to log to
+   */
+  logger(std::string name, std::ostream& stream);
+
+  /**
+   * @brief Construct a new logger object
+   *
+   * @param name The name of the logger
+   * @param sinks The sinks to log to
+   *
+   * Note that we must use a vector because initializer_lists are not flexible
+   * enough to support programmatic construction in callers, and an
+   * iterator-based API would require templating and thus exposing spdlog
+   * types.
+   */
+  logger(std::string name, std::vector<std::shared_ptr<sink>> sinks);
+
+  /**
+   * @brief Destroy the logger object
+   */
+  ~logger();
+
+  /**
+   * @brief Log a message at the specified level.
+   *
+   * This is the core logging routine that dispatches to spdlog.
+   *
+   * @param lvl The log level
+   * @param message The message to log
+   */
+  void log(level_enum lvl, std::string const& message);
+
+  /**
+   * @brief Add a sink to the logger.
+   *
+   * @param sink The sink to add
+   */
+  void add_sink(spdlog::sink_ptr sink);
+
+  /**
+   * @brief Remove a sink from the logger.
+   *
+   * @param sink The sink to remove
+   */
+  void remove_sink(spdlog::sink_ptr sink);
```

Comment:
I mainly defined these functions rather than exposing a `sinks` function directly (like spdlog) to ensure that the API would be consistent whether `SUPPORTS_LOGGING` is on or off. If `sinks` were exposed directly we'd have to jump through more hoops to ensure that the vector behaved reasonably in cases where logging is disabled (should there be sinks? what attributes do they possess? etc).